<a href="https://colab.research.google.com/github/vkusnezh/Projects-Portfolio/blob/main/Fake_news_detector.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

The goal of this project is to create a sand-box to play with building a models to classify a news as REAL or FAKE.

# News definition:


# Fake news definition

Fake news - False information that is broadcast or published as news for fraudulent or politically motivated purposes [2].


# Project

In [ ]:
import numpy as np
import pandas as pd
import itertools
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import PassiveAggressiveClassifier
from sklearn.metrics import accuracy_score, confusion_matrix

# New Section

In [ ]:
#Read the data from google drive
url='https://drive.google.com/file/d/1g1DS0TbDYWKVORIbC3XJJ6Do73KrkQ9b/view?usp=sharing'
url_c='https://drive.google.com/uc?export=download&id='+url.split('/')[-2]
df=pd.read_csv(url_c)

#Get head

#df.shape
df.head()

Unnamed: 0  ... label
0        8476  ...  FAKE
1       10294  ...  FAKE
2        3608  ...  REAL
3       10142  ...  FAKE
4         875  ...  REAL

[5 rows x 4 columns]

In [ ]:
#DataFlair - get the labels from the DataFrame
labels=df.label
labels.head()

0    FAKE
1    FAKE
2    REAL
3    FAKE
4    REAL
Name: label, dtype: object

In [ ]:
#Split the dataset into training and testing sets
x_train,x_test,y_train,y_test=train_test_split(df['text'], labels, test_size=0.3, random_state=10)

Initialization of a TfidfVectorizer with stop words from the English language and a maximum document frequency of 0.7 (terms with a higher document frequency will be discarded). 
TfidfVectorizer turns a collection of raw documents into a matrix of TF-IDF features.

Fiting and transformation of the vectorizer on the train set, and transformation of the vectorizer on the test set.

In [ ]:
#DataFlair - Initialize a TfidfVectorizer
tfidf_vectorizer=TfidfVectorizer(stop_words='english', max_df=0.7)

#DataFlair - Fit and transform train set, transform test set
tfidf_train=tfidf_vectorizer.fit_transform(x_train) 
tfidf_test=tfidf_vectorizer.transform(x_test)

Initialization of a PassiveAggressiveClassifier, and fit it on tfidf_train and y_train.

Then, we’ll predict on the test set from the TfidfVectorizer and calculate the accuracy with accuracy_score() from sklearn.metrics.

In [ ]:
#DataFlair - Initialize a PassiveAggressiveClassifier
pac=PassiveAggressiveClassifier(max_iter=30)
pac.fit(tfidf_train,y_train)
#DataFlair - Predict on the test set and calculate accuracy
y_pred=pac.predict(tfidf_test)
score=accuracy_score(y_test,y_pred)
print(f'Accuracy: {round(score*100,3)}%')

Accuracy: 94.214%


Accuracy is 94.2% with this model. Finally, let’s print out a confusion matrix to gain insight into the number of false and true negatives and positives.

In [ ]:
#DataFlair - Build confusion matrix
confusion_matrix(y_test,y_pred, labels=['FAKE','REAL'])

array([[874,  46],
       [ 64, 917]])

So with this model, we have 874 true positives, 917 true negatives, 46 false positives, and 64 false negatives.

# Conclusions

# References



1.   https://data-flair.training/blogs/advanced-python-project-detecting-fake-news/
2.   https://www.lexico.com/definition/fake_news
3.   sdsf
  


